src/main.cc

```diff
@@ -2043,6 +2070,10 @@ watch_child(const CommandLine &masterCommand)
             syslog(LOG_NOTICE, "Squid Parent: unknown child process %d exited", pid);
 
         masterCheckAndBroadcastSignals();
+        if (restarting_kids) {
+            TheKids.forgetAllFailures();
```

Comment:
> Why do we have to delay this forgetAllFailures() 

I was first thinking that since restarting itself is a continuous action, we should have a 'state' variable for this action (like, e.g., reconfiguring or shutting_down). However, managing this variable (and reset it when required) is probably a useless overhead, so I removed it. It is possible to revive hopeless kids immediately, of course.

> but we also should be ignoring a single failure

I added a new Kids:forgetHopelessFailures() to address this concern: restarting should not clear error counters of not-yet hopeless processes.

Reply:
> I was first thinking that since restarting itself is a continuous action, we should have a 'state' variable for this action (like, e.g., reconfiguring or shutting_down). However, managing this variable (and reset it when required) is probably a useless overhead, so I removed it.

Ideally, we want to ignore the kid error that occurred due to that kid restart and honor any errors that occurred after restart. I do not know for sure, but I suspect that ideal is impossible to reach/guarantee with reasonable effort, but we can probably get close in most common situations. To do that, we may need a variable (or even a Kid data member), but I do not know that either.

Also, we need to think about the admin intent here: Why are they restarting Squid kids without shutting down the master process? What are they trying to gain by that?

* Without SMP, I can imagine very few advantages of such a restart, but I am not a Squid admin, and the feature was added way before SMP days so there must be a reason for its existence that I am missing.
* With SMP, I can imagine supporting a _graceful_ restart where Squid continues to serve traffic during restart while individual kids are restarted sequentially, in the right order. I do not know whether we can support such restarts without significant kid modifications (which are out of this issue scope), but it is something to keep in mind when designing your changes. And if restarting SMP kids breaks listening socket and/or rock cache_dir registrations/sycnrhonization, then we should warn folks about it -- it may not actually make sense to restore this lost feature in SMP world without investing in graceful restart handling! Or perhaps we should resrtart only workers (or workers and diskers but not the coordinator) to make this feature work.

I suggest asking the admin who complained that the feature is broken on Bugzilla -- why are they using "squid -k restart" instead of "service squid restart" (or equivalent)? What are they trying to optimize or prevent by using that feature? One admin/use case may not be good enough to drive all our decisions, but it may still be a useful data point.

> restarting should not clear error counters of not-yet hopeless processes

Why should not it?

I think it is reasonable to start with a clean slate after an admin-triggered restart. We are already reviving hopeless kids and they were in a  much worse shape. Remembering kid errors at all is only needed to detect hopeless kids IIRC. If we are OK with starting previously hopeless kids, we should be OK with resetting the counter for running kids. If a running kid was at the crisp of being declared hopeless, then we will re-learn that fast enough in most cases -- clearing that counter should not delay hopeless kids detection by much in most cases, right?

Forgetting all errors also makes/keeps the code simpler.

If there are no specific good reasons why we should preserve pre-restart errors, I would not preserve them.

Reply:
> Forgetting all errors also makes/keeps the code simpler.

Done at bc2581a8.